src/SAWScript/MGU.hs

```diff
@@ -554,36 +646,64 @@ inferField m (n,e) = do
   (e',t) <- inferE (m,e)
   return ((n,e'),(n,t))
 
-bindSchema :: Located Name -> Schema -> TI a -> TI a
-bindSchema n s m = TI $ local (\ro -> ro { typeEnv = M.insert n s $ typeEnv ro })
-  $ unTI m
+-- wrap m with a type for x
+withVar :: Located Name -> Schema -> TI a -> TI a
+withVar x s m =
+  TI $ local (\ro -> ro { varEnv = M.insert x s $ varEnv ro }) $ unTI m
 
-bindSchemas :: [(Located Name, Schema)] -> TI a -> TI a
-bindSchemas bs m = foldr (uncurry bindSchema) m bs
+-- wrap m with types for a list of vars
+withVars :: [(Located Name, Schema)] -> TI a -> TI a
+withVars bindings m = foldr (uncurry withVar) m bindings
 
-bindPattern :: Pattern -> TI a -> TI a
-bindPattern pat = bindSchemas bs
-  where bs = [ (x, tMono t) | (x, Just t) <- patternBindings pat ]
-
-bindPatternSchema :: Pattern -> Schema -> TI a -> TI a
-bindPatternSchema pat s@(Forall vs t) m =
+-- wrap m with types for all the vars in a pattern
```

Comment:
An arbitrary action in ```TI```. Basically it'll be the checking for everything following the bind/declaration that added the variables.

I updated the comments to be more explicit in bf86a7f58c25f50c872b7d3455033fa784d7192e

Reply:
Sorry, I should have been more explicit here. I understand that you're describing a `TI` action here, but you are also naming a variable `m` that simply does not exist anywhere in the implementation of the function. Compare the implementation of `withPattern`:

```hs
-- wrap m with types for all the vars in a pattern
-- ...
withPattern :: Pattern -> TI a -> TI a
withPattern pat = withVars bindings
  where bindings = [ (x, tMono t) | (x, Just t) <- patternBindings pat ]
```

With another function, like `withVar`:

```hs
-- wrap m with a type for x
withVar :: Located Name -> Schema -> TI a -> TI a
withVar x s m =
  TI $ local (\ro -> ro { varEnv = M.insert x s $ varEnv ro }) $ unTI m
```

`withVar` actually has an argument named `m`, so its comment about "wrap `m`" makes sense. `withPattern`, on the other hand, doesn't have an argument named `m` in its implementation, which leaves readers to wonder why the comment bothers to name `m` in the first place. (The name `m` isn't used anywhere in the rest of the `withPattern`'s comment, either.)

I would favor just saying something like "wrap the action" instead of "wrap `m`" to avoid the opportunity for confusion here.

Reply:
oh ffs, I didn't even notice that that one had been eta-reduced. I'll fix that. Gratuitous eta-reduction is not helpful for code legibility... I put an ```m``` back in, not having it just confuses things.